pkg/readiness/ready_tracker.go

```diff
@@ -533,169 +578,224 @@ func (t *Tracker) collectInvalidExpectations(ctx context.Context) {
 	}
 }
 
-func (t *Tracker) trackAssignMetadata(ctx context.Context) error {
+func (t *Tracker) trackAssignMetadata(ctx context.Context, errChan chan<- error) {
+	hadError := false
 	defer func() {
-		t.assignMetadata.ExpectationsDone()
-		log.V(logging.DebugLevel).Info("AssignMetadata expectations populated")
-
-		_ = t.constraintTrackers.Wait()
+		// If we are ignoring errors when tracking expecations, we need to set expectations to done to prevent readiness tracker never being satisfied
+		if !t.crashOnFailure || !hadError {
+			t.assignMetadata.ExpectationsDone()
+			log.V(logging.DebugLevel).Info("AssignMetadata expectations populated")
+		}
 	}()
 
 	if !t.mutationEnabled {
-		return nil
+		return
 	}
 
 	assignMetadataList := &mutationv1.AssignMetadataList{}
-	lister := retryLister(t.lister, retryAll)
+	retryPred := retryAll
```

Comment:
nit: `retryPredicate` here and elsewhere -- `retryPred` is not that much shorter and `Predicate` is kind of a niche word